executable/microsoft_pe.ksy

```diff
@@ -408,3 +413,144 @@ types:
         size: length - 8
         doc: Contains a certificate, such as an Authenticode signature.
     doc-ref: 'https://docs.microsoft.com/en-us/windows/desktop/debug/pe-format#the-attribute-certificate-table-image-only'
+
+  resource_directory_table:
+    doc-ref: 'https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#resource-directory-table'
+    params:
+      - id: depth
+        type: u4
+      - id: section_file_offset
+        type: u4
+      - id: section_virtual_address
+        type: u4
+    seq:
+      - id: reserved
+        -orig-id: characteristics
+        type: u4
+        doc: Reserved for future use, should be set to zero
+      - id: timestamp
+        type: u4
+        -orig-id: TimeDateStamp
+      - id: version
+        type: version_u2
+      - id: num_named_entries
+        -orig-id: NumberOfNamedEntries
+        type: u2
+      - id: num_id_entries
+        -orig-id: NumberOfIdEntries
+        type: u2
+      - id: named_entries
+        type: resource_directory_entry(depth)
+        repeat: expr
+        repeat-expr: num_named_entries
+      - id: id_entries
+        type: resource_directory_entry(depth)
+        repeat: expr
+        repeat-expr: num_id_entries
+
+  resource_directory_entry:
+    doc-ref: 'https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#resource-directory-entries'
+    params:
+      - id: depth
+        type: u4
+    seq:
+      - id: resource_type
+        type: u4
+        enum: enum_resource_type
+        if: depth == 0
+      - id: ofs_name
+        type: b31le
+        if: depth != 0
+      - id: is_name
+        type: b1be
+        if: depth != 0
+        doc: Whether ofs_name is an id or points to a directory-string
+      - id: ofs_data_entry
+        type: b31le
+      - id: has_child
+        type: b1be
```

Comment:
LE and BE bit-sized fields cannot follow on an unaligned bit position. Right now, it has **undefined behavior**, and KSC will respond to it with a compile-time error in the future. See https://doc.kaitai.io/user_guide.html#bit-endian, the _Important_ section:

> _Big-endian_ and _little-endian_ bit integers can follow **only on a byte boundary**. They can’t share the same byte. Joining them on an unaligned bit position is _undefined behavior_, and KSC will be throwing a compile-time error in the future if it detects such situation.
>
> For example, this is **illegal**: (...)

```suggestion
      - id: ofs_data_entry
        type: b31le
      - id: has_child
        type: b1be
```

---

I suggest placing 

```ksy
meta:
  bit-endian: le
```

on the very top of the file, probably below `endian` , because that's not gonna change throughout the file. Then simply omit the `le` suffixes after each bit type:

```suggestion
      - id: ofs_data_entry
        type: b31
      - id: has_child
        type: b1
```

etc.

Reply:
Thanks for the thorough review and explanation.

I admit I was a bit confused how the bit-endianness worked. I somehow expected `b1le` to actually returns `0` on `80` (i.e. take the 4th byte, interpret the bit in little-endian, returns the first bit which is 0 in the least significant bit).